3.0/m_roleplay.cpp

```diff
@@ -0,0 +1,704 @@
+	CmdResult Handle(User* user, const Params& parameters) CXX11_OVERRIDE
+	{
+		if(!user->HasPrivPermission("channels/roleplay"))
+		{
+			user->WriteNumeric(ERR_NOPRIVILEGES, "Permission Denied - You do not have the required operator privileges");
+			return CMD_FAILURE;
+		}
+
+		return CommandBaseRoleplay::Handle(user, parameters);
+	}
+};
+
+class CommandFAction : public CommandBaseRoleplay
+{
+protected:
+	std::string GetSource(const Params& parameters)
+	{
+		return (ServerInstance->IsNick(parameters[1]) ? parameters[1] : "");
+	}
+
+	std::string GetMessage(const Params& parameters)
+	{
+		return MakeAction(parameters[2]);
+	}
+
+public:
+	CommandFAction(Module* Creator, RoleplayMode& mode)
+		: CommandBaseRoleplay(Creator, "FACTION", 3, mode)
+	{
+		syntax = "<channel> <nickname> :<message>";
+		flags_needed = 'o';
+	}
+
+	CmdResult Handle(User* user, const Params& parameters) CXX11_OVERRIDE
+	{
+		if(!user->HasPrivPermission("channels/roleplay"))
+		{
+			user->WriteNumeric(ERR_NOPRIVILEGES, "Permission Denied - You do not have the required operator privileges");
+			return CMD_FAILURE;
+		}
+
+		return CommandBaseRoleplay::Handle(user, parameters);
+	}
+};
+
+class CommandNPC : public CommandBaseRoleplay
+{
+protected:
+	std::string GetSource(const Params& parameters)
+	{
+		if(!ServerInstance->IsNick(parameters[1]))
+			return "";
+
+		return MakeUnderline(parameters[1]);
+	}
+
+	std::string GetMessage(const Params& parameters)
+	{
+		return parameters[2];
+	}
+
+public:
+	CommandNPC(Module* Creator, RoleplayMode& mode)
+		: CommandBaseRoleplay(Creator, "NPC", 3, mode)
+	{
+		syntax = "<channel> <nickname> :<message>";
+	}
+};
+
+class CommandNPCA : public CommandBaseRoleplay
+{
+protected:
+	std::string GetSource(const Params& parameters)
+	{
+		if(!ServerInstance->IsNick(parameters[1]))
+			return "";
+
+		return MakeUnderline(parameters[1]);
+	}
+
+	std::string GetMessage(const Params& parameters)
+	{
+		return MakeAction(parameters[2]);
+	}
+
+public:
+	CommandNPCA(Module* Creator, RoleplayMode& mode)
+		: CommandBaseRoleplay(Creator, "NPCA", 3, mode)
+	{
+		syntax = "<channel> <nickname> :<message>";
+	}
+};
+
+class ModuleRoleplay : public Module
+{
+	RoleplayMode roleplaymode;
+	CommandScene cscene;
+	CommandSceneA cscenea;
+	CommandAmbiance cambiance;
+	CommandNarrator cnarrator;
+	CommandNarratorA cnarratora;
+	CommandFSay cfsay;
+	CommandFAction cfaction;
+	CommandNPC cnpc;
+	CommandNPCA cnpca;
+
+public:
+	ModuleRoleplay()
+		: roleplaymode(this)
+		, cscene(this, roleplaymode)
+		, cscenea(this, roleplaymode)
+		, cambiance(this, roleplaymode)
+		, cnarrator(this, roleplaymode)
+		, cnarratora(this, roleplaymode)
+		, cfsay(this, roleplaymode)
+		, cfaction(this, roleplaymode)
+		, cnpc(this, roleplaymode)
+		, cnpca(this, roleplaymode)
+	{
+	}
+
+	void ReadConfig(ConfigStatus& status) CXX11_OVERRIDE
+	{
+		ConfigTag* tag = ServerInstance->Config->ConfValue("roleplay");
+		bool need_mode = tag->getBool("needchanmode", true);
```

Comment:
This will not enable the mode later.

Instead of reading this in `ReadConfig` you can just do `mh->GetId() != ModeParser::MODEID_MAX` to see if the mode is registered.

Reply:
I think I fixed this.